### PR TITLE
[CLANG]Add Scalable vectors for mfloat8_t

### PR DESCRIPTION
**Pull request by CarolineConcatto**
This patch adds these new vector sizes for sve:
    svmfloat8_t

According to the ARM ACLE PR#323[1].

[1] ARM-software/acle#323

### CODE REVIEW

File: clang/lib/AST/ASTContext.cpp [hunk 1]
```diff
@@ -4339,7 +4339,6 @@ ASTContext::getBuiltinVectorTypeInfo(const BuiltinType *Ty) const {
   switch (Ty->getKind()) {
   default:
     llvm_unreachable("Unsupported builtin vector type");
-
```
Comment (paulwalker-arm) [inline clang/lib/AST/ASTContext.cpp:4342]: Please can we keep these blank lines.  With all the continuation markers they help partition the related defines and includes.

File: clang/lib/AST/ASTContext.cpp [hunk 2]
```diff
@@ -4354,12 +4353,16 @@ ASTContext::getBuiltinVectorTypeInfo(const BuiltinType *Ty) const {
...
+                               ElBits, NF)                                     \
+  case BuiltinType::Id:                                                        \
+    return {getIntTypeForBitwidth(ElBits, false),                              \
+            llvm::ElementCount::getScalable(NumEls), NF};
 #define SVE_PREDICATE_TYPE_ALL(Name, MangledName, Id, SingletonId, NumEls, NF) \
   case BuiltinType::Id:                                                        \
     return {BoolTy, llvm::ElementCount::getScalable(NumEls), NF};
 #define SVE_OPAQUE_TYPE(Name, MangledName, Id, SingletonId)
 #include "clang/Basic/AArch64SVEACLETypes.def"
-
```
Comment (paulwalker-arm) [inline clang/lib/AST/ASTContext.cpp:4362]: As above.

File: clang/include/clang/Basic/AArch64SVEACLETypes.def [hunk 3]
```diff
@@ -69,6 +69,11 @@
 
 #ifndef SVE_VECTOR_TYPE_BFLOAT
 #define SVE_VECTOR_TYPE_BFLOAT(Name, MangledName, Id, SingletonId, NumEls, ElBits, NF) \
+  SVE_VECTOR_TYPE_DETAILS(Name, MangledName, Id, SingletonId, NumEls, ElBits, NF, true, false, true)
+#endif
+
+#ifndef SVE_VECTOR_TYPE_MFLOAT
+#define SVE_VECTOR_TYPE_MFLOAT(Name, MangledName, Id, SingletonId, NumEls, ElBits, NF) \
```
Comment (paulwalker-arm) [inline clang/include/clang/Basic/AArch64SVEACLETypes.def:76]: Is this necessary.  Looking at the uses I think `SVE_VECTOR_TYPE_INT` would work?

File: clang/utils/TableGen/SveEmitter.cpp [hunk 4]
```diff
@@ -87,6 +87,10 @@ class SVEType {
   bool isDefault() const { return DefaultType; }
   bool isFloat() const { return Float && !BFloat; }
   bool isBFloat() const { return BFloat && !Float; }
+  bool isMFloat() const {
+    return MFloat && !BFloat && !Float;
+    ;
```
Comment (paulwalker-arm) [inline clang/utils/TableGen/SveEmitter.cpp:92]: Stray `;`

File: clang/utils/TableGen/SveEmitter.cpp [hunk 5]
```diff
@@ -87,6 +87,10 @@ class SVEType {
   bool isDefault() const { return DefaultType; }
   bool isFloat() const { return Float && !BFloat; }
   bool isBFloat() const { return BFloat && !Float; }
```
Comment (paulwalker-arm) [inline clang/utils/TableGen/SveEmitter.cpp:89]: To be consistent I think these need updating to include `&& !MFloat`?

There's also an "isFloatingPoint()" method that probably needs `|| MFloat`.

File: clang/utils/TableGen/SveEmitter.cpp [hunk 6]
```diff
@@ -454,6 +458,8 @@ std::string SVEType::builtin_str() const {
   else if (isBFloat()) {
     assert(ElementBitwidth == 16 && "Not a valid BFloat.");
     S += "y";
+  } else if (isMFloat()) {
+    S += "m";
```
Comment (paulwalker-arm) [inline clang/utils/TableGen/SveEmitter.cpp:462]: Perhaps worth adding `assert(ElementBitwidth == 8 && "Not a valid MFloat.")`?

**Review by paulwalker-arm — COMMENTED** [review]
Not sure if there's enough support to update all of them but, but many of the following should be extended to cover the new types:

clang/test/CodeGenCXX/aarch64-sve-vector-init.cpp
clang/test/CodeGenCXX/aarch64-mangle-sve-fixed-vectors.cpp
clang/test/CodeGenCXX/aarch64-mangle-sve-vectors.cpp
clang/test/CodeGenCXX/aarch64-sve-fixedtypeinfo.cpp
clang/test/CodeGenCXX/aarch64-sve-typeinfo.cpp
clang/test/AST/ast-dump-aarch64-sve-types.c
clang/test/CodeGen/aarch64-sve.c
clang/test/CodeGenObjC/aarch64-sve-types.m
clang/test/Sema/aarch64-sve-types.c
clang/test/Sema/attr-arm-sve-vector-bits.c
clang/test/SemaObjC/aarch64-sve-types.m

Is it worth adding the tuple types as well? I'm happy for that to be a separate PR but I figure you'll be touching the same files so it might be easier done together.  Your choice.

**Comment by CarolineConcatto** [comment]
Hi @paulwalker-arm,
I believe I addressed all the comments in the code. 

I updated almost all of them to have a test for the SVMfloat8_t.
The ones that are not updated are the ones that test FIXED SVE types. 
The mangling needs scalar type support, BuiltinType::MFloat8,  in order to NOT mangle to SVUint8 (CXXNameMangler::mangleAArch64FixedSveVectorType(const VectorType *T))
I believe I can add support for that when the scalar type is merged.
About the tuple types.
If you dont mind I would rather address that in another patch.

File: clang/lib/AST/ASTContext.cpp [hunk 7]
```diff
@@ -4419,6 +4419,13 @@ QualType ASTContext::getScalableVectorType(QualType EltTy, unsigned NumElts,
       EltTySize == ElBits && NumElts == (NumEls * NF) && NumFields == 1) {     \
     return SingletonId;                                                        \
   }
+#define SVE_VECTOR_TYPE_MFLOAT(Name, MangledName, Id, SingletonId, NumEls,     \
+                               ElBits, NF)                                     \
+  if (EltTy->hasIntegerRepresentation() && !EltTy->isBooleanType() &&          \
+      !EltTy->hasSignedIntegerRepresentation() && EltTySize == ElBits &&       \
+      NumElts == (NumEls * NF) && NumFields == 1) {                            \
+    return SingletonId;                                                        \
+  }
```
Comment (paulwalker-arm) [inline clang/lib/AST/ASTContext.cpp:4428]: This is no longer required?

File: clang/include/clang/Basic/AArch64SVEACLETypes.def [hunk 8]
```diff
@@ -69,7 +69,7 @@
 
 #ifndef SVE_VECTOR_TYPE_BFLOAT
 #define SVE_VECTOR_TYPE_BFLOAT(Name, MangledName, Id, SingletonId, NumEls, ElBits, NF) \
-  SVE_VECTOR_TYPE_DETAILS(Name, MangledName, Id, SingletonId, NumEls, ElBits, NF, false, false, true)
+  SVE_VECTOR_TYPE_DETAILS(Name, MangledName, Id, SingletonId, NumEls, ElBits, NF, true, false, true)
```
Comment (paulwalker-arm) [inline clang/include/clang/Basic/AArch64SVEACLETypes.def:72]: Is this change necessary?

File: clang/lib/AST/ASTContext.cpp (hunk 7 above)
Comment (CarolineConcatto) [reply clang/lib/AST/ASTContext.cpp]: I missed that when I removed the MFLOAT.

File: clang/include/clang/Basic/AArch64SVEACLETypes.def (hunk 8 above)
Comment (CarolineConcatto) [reply clang/include/clang/Basic/AArch64SVEACLETypes.def]: I missed that when I removed the MFLOAT.

File: clang/utils/TableGen/SveEmitter.cpp (hunk 6 above)
Comment (CarolineConcatto) [reply clang/utils/TableGen/SveEmitter.cpp:462]: Done

**Review by paulwalker-arm — APPROVED** [review]
